src/invidious/connection/pool.cr

```diff
@@ -0,0 +1,116 @@
+module Invidious::ConnectionPool
+  struct Pool
+    property url : URI
+    property pool : DB::Pool(HTTP::Client)
+
+    def initialize(
+      url : URI,
+      *,
+      max_capacity : Int32 = 5,
+      idle_capacity : Int32? = nil,
+      timeout : Float64 = 5.0
+    )
+      if idle_capacity.nil?
+        idle_capacity = max_capacity
+      end
+
+      @url = url
+
+      options = DB::Pool::Options.new(
+        initial_pool_size: 0,
+        max_pool_size: max_capacity,
+        max_idle_pool_size: idle_capacity,
+        checkout_timeout: timeout
+      )
+
+      @pool = DB::Pool(HTTP::Client).new(options) do
+        next make_client(url, force_resolve: true)
+      end
+    end
+
+    {% for method in %w[get post put patch delete head options] %}
+      def {{method.id}}(*args, **kwargs, &)
+        self.client do | client |
+          client.{{method.id}}(*args, **kwargs) do | response |
+
+            result = yield response
+            return result
+
+          ensure
+            response.body_io?.try &. skip_to_end
+          end
+        end
+      end
+
+      def {{method.id}}(*args, **kwargs)
+        {{method.id}}(*args, **kwargs) do | response |
+          return response
+        ensure
+          response.body_io?.try &. skip_to_end
+        end
+      end
+    {% end %}
+
+    # Checks out a client in the pool
+    private def client(&)
+      # If a client has been deleted from the pool
+      # we won't try to release it
+      client_exists_in_pool = true
+
+      http_client = pool.checkout
+
+      # When the HTTP::Client connection is closed, the automatic reconnection
+      # feature will create a new IO to connect to the server with
+      #
+      # This new TCP IO will be a direct connection to the server and will not go
+      # through the proxy. As such we'll need to reinitialize the proxy connection
+
+      http_client.proxy = make_configured_http_proxy_client() if CONFIG.http_proxy
+
+      response = yield http_client
+    rescue ex : DB::PoolTimeout
+      # Failed to checkout a client
+      raise ConnectionPool::Error.new(ex.message, cause: ex)
```

Comment:
Is it even useful for us to add the cause of the ConnectionPool exception here when we know that the original error is a `DB::PoolTimeout`? All it does really is produce a lengthy backtrace that can potentially confuse users